### DISCUSSION TITLE
The decompile results looks wrong, why the refinded Pcode is not complete for this function

### DISCUSSION THREAD

**Discussion by jackniu1234** [NationalSecurityAgency/ghidra#7078] [Q&A]
I'm trying to use Ghidra's refined pcode to do some data-flow analysis, however, when I dumped the refined pcodes of the below function, it seems that the refined pcodes are not complete. Is there anyway to use some decomiledOptions to fix it?

Here is the function written in C, and compiled into an ELF file with ARM:LE:32:v8 using arm-none-eabi-gcc.

```c
int func(int* p, int sz) {
    p = p + 1;
    if (sz > 10) 
        p = p + 1;
    else
        p = p - 1;
    return *p;
}
```

The dumped refined pcodes look like:
![image](https://github.com/user-attachments/assets/c5305606-9826-4ce2-8a4c-bd76af80f984)

The decompiled function in "Decompile window" of GUI also looks wrong:
![image](https://github.com/user-attachments/assets/0665ff61-88a7-496f-89da-4ce60c892eb8)

**Comment by ghidracadabra** [comment]
Can you elaborate on what looks wrong in the decompiled code?  The decompiler can perform its own simplifications which result in different (but equivalent) C code than what was originally compiled.  Optimizations applied by the compiler can also have a role.

The unused local `sz_local` is most likely a result of the DWARF analyzer applying information to a local variable that the decompiler optimized away.

**Comment by ernieIzde8ski** [comment]
These functions might not look the same, but they're logically equivalent, with the compiler/decompiler applying some degree of obfuscation:

<table>

  <tr>

  <th> Code </th>
  <th> Diff </th>
  <th> Justification </th>

  </tr>

  <tr>

  <td>

```c
int func(int* p, int sz) {
  p = p + 1;
  if (sz > 10)
    p = p + 1;
  else
    p = p - 1;
  return *p;
}
```

  </td>

  <td> N/A </td>

  <td> Given </td>

  </tr>

  <tr>

  <td>

```c
int func(int* p, int sz) {
  p = p;
  if (sz > 10)
    p = p + 2;
  else
    p = p;
  return *p;
}
```

  </td>

  <td>
  <details>
  <summary> Diff </summary>

```diff
  int func(int* p, int sz) {
-     p = p + 1;
+     p = p;
      if (sz > 10)
-         p = p + 1;
+         p = p + 2;
      else
-         p = p - 1;
+         p = p;
      return *p;
  }
```

  </details>

  </td>

  <td> Move <code>p + 1</code> out of the first line. </td>

  </tr>

  <tr>

  <td>

```c
int func(int* p, int sz) {
  if (sz > 10) {
    p = p + 2;
  }
  return *p;
}
```

  </td>

  <td>

  <details>
  <summary> Diff </summary>

```diff
  int func(int* p, int sz) {
-     p = p;
      if (sz > 10) {
          p = p + 2;
      }
-     else
-         p = p;
      return *p;
  }
```

  </details>

  </td>

  <td> Remove identity equalities. </td>

  </tr>

  <tr>

  <td>

```c
  int func(int* p, int sz) {
    if (10 < sz) {
        p = p + 2;
    }
    return *p;
}
```

  </td>

  <td>

  <details>
  <summary> Diff </summary>

```diff
  int func(int* p, int sz) {
-     if (sz > 10) {
+     if (10 < sz) {
          p = p + 2;
      }
      return *p;
  }
```

  </details>
  </td>

  <td> Flip the ordering of the equality. </td>

  </tr>

  <tr>

  <td>

```c
int func(int* p, int sz) {
  int sz_local;
  int *p_local;

  p_local = p;
  if (10 < sz) {
      p_local = p + 2;
  }
  return *p_local;
}
```

  </td>

  <td>

  <details>
  <summary> Diff </summary>

```diff
  int func(int* p, int sz) {
+   int sz_local;
+   int *p_local;
+ 
+   p_local = p;
    if (10 < sz) {
-       p = p + 2;
+       p_local = p + 2;
    }
    return *p_local;
  }
```

  </details>
  </td>

  <td>
Introduce locals. Presumably, sz_local is an artifact like the other user says.
  </td>

  </tr>

</table>
